### ISSUE TITLE
State changing, but not re-rendering

### ISSUE THREAD

**Issue by MMasterson** [xavierLowmiller/AppStorage#5]
I have a simple MVVM structure, where a toggle in a view changes the state of a property in the view model. The property changes and saves to user defaults  fine, but the ObservedObject doesn't notice a state change there for does not re-render. 

Sample code below

```
class SampleViewModel: ObservableObject {
    @AppStorageCompat("includePhoto") var state: Bool = false
    
    func someFunc() {
        return state.description
    }
    
}

struct TheView: View {
    
    
    @ObservedObject var vm: SampleViewModel = SampleViewModel()
    
    var body: some View {
        VStack{
            Text(self.vm.someFunc())
            Toggle(left, isOn: self.vm.$state).labelsHidden()
            
        }
}
```

**Comment by xavierLowmiller** [comment]
Hi @MMasterson,

Sorry, but this won't work.

TLDR: [@AppStorage](https://developer.apple.com/documentation/swiftui/appstorage) does [not work](https://stackoverflow.com/questions/63448377/how-to-update-swiftui-view-after-appstorage-in-a-separate-struct-gets-updated) outside a `View` struct, so this compatibility wrapper does not, too.

Longer answer: To make an `ObservableObject` update, there are two ways:
1. Mark properties as `@Published`, but property wrapper composition [isn't trivial](https://nshipster.com/propertywrapper/#property-wrappers-are-difficult-to-compose):
```swift
@Published
@AppStorageCompat("includePhoto")
var state: Bool = false
```
2. Update the `objectWillChange` property manually:
```swift
@AppStorageCompat("includePhoto") var state: Bool = false {
    willSet {
        objectWillChange.send()
    }
}
```
But this won't work because `willSet`/`didSet` doesn't mean anything to property wrappers (see point 4 [here](https://www.hackingwithswift.com/articles/224/common-swiftui-mistakes-and-how-to-fix-them)).

**Comment by MMasterson** [comment]
@xavierLowmiller Thanks for the info and the links. This is my first SwiftUI project, so I appreciate the detailed answer

I've found a work around of updating a state property whenever I update the `AppStorageCompact` props, a little cumbersome, but it'll work until I find a more elegant solution.

**Comment by aheze** [comment]
@xavierLowmiller thanks for making this library. The `willSet` workaround works for me though, is it not supposed to?

**Comment by xavierLowmiller** [comment]
Interesting! Judging by the [StackOverflow answer](https://stackoverflow.com/a/63492773/4239752), it looks like iOS 14.5 changed the behavior here.
I haven't looked at this in a while, maybe it's just supported now. I'll take a look and update the thread.

**Comment by aheze** [comment]
Thanks! Actually I did some more testing and it seems like:

- When you directly set the variable, `willSet` is called.
```swift
state = true
```

- When you set the variable via a Binding, `willSet` is **NOT** called.
```swift
Toggle("", isOn: $state) /// `willSet` won't be called!
```

My current workaround is:

1. Use `willSet` to handle directly setting the variable
2. Add a `valueChanged` closure and call it whenever `wrappedValue` changes:
```diff
@frozen @propertyWrapper public struct AppStorageCompat<Value>: DynamicProperty {
    @ObservedObject private var _value: Storage<Value>
    private let saveValue: (Value) -> Void
+    var valueChanged: (() -> Void)?

    private init(value: Value, store: UserDefaults, key: String, transform: @escaping (Any?) -> Value?, saveValue: @escaping (Value) -> Void) {
        self._value = Storage(value: value, store: store, key: key, transform: transform)
        self.saveValue = saveValue
    }

    public var wrappedValue: Value {
        get {
            _value.value
        }
        nonmutating set {
            saveValue(newValue)
            _value.value = newValue

+            if let valueChanged = valueChanged {
+                valueChanged()
+            } else {
+                assert(false, "`valueChanged` not set for \(self)")
+            }
        }
    }
```

3. Set `valueChanged` to call `objectWillChange.send()`.
```swift
class SampleViewModel: ObservableObject {
    @AppStorageCompat("includePhoto") var state: Bool = false {
        willSet {
            self.objectWillChange.send()
        }
    }

    init() {
        _state.configureValueChanged(with: self)
    }
}

extension @AppStorageCompat {
    /// listen to value changed
    mutating func configureValueChanged(with model: SampleViewModel) {
        valueChanged = { [weak model] in
            model?.objectWillChange.send()
        }
    }
}
```

This is extremely verbose and not ideal at all, but it works. Unfortunately it seems like you can't pass `self` into a property wrapper — otherwise I was thinking of something like `@AppStorageCompat("includePhoto", self)` and call `self.objectWillChange.send()` inside the property wrapper. Any ideas? Thanks again for taking a look!
